Add initial WebGPU env validation

Kangz opened the pull request:
@dneto0 PTAL!

This adds support for the "webgpu0" = "WIP WebGPU" environment and adds an initial check that disallows usage of OpUndef (+ a test).

Also some minor cleanups / warning fixes.

Kangz commented on source/validate_instruction.cpp:
```diff
@@ -205,6 +205,12 @@ spv_result_t ReservedCheck(ValidationState_t& _,
       return _.diag(SPV_ERROR_INVALID_BINARY)
              << "Invalid Opcode name 'Op" << inst_desc->name << "'";
     }
+    case SpvOpUndef:
```
This function is called CheckReserved, should the name be updated to reflect that we also do env-related checks?

dneto0 replied:
Also the comment on the method would need updating.

The alternative is to make a whole new function.

Kangz replied:
Chose to add a new EnvironmentCheck function.

dneto0 commented on include/spirv-tools/libspirv.h:
```diff
@@ -412,6 +412,7 @@ typedef enum {
   SPV_ENV_OPENCL_EMBEDDED_2_2,  // OpenCL Embedded Profile 2.2 latest revision.
   SPV_ENV_UNIVERSAL_1_3,  // SPIR-V 1.3 latest revision, no other restrictions.
   SPV_ENV_VULKAN_1_1,     // Vulkan 1.1 latest revision.
+  SPV_ENV_WEBGPU_0, // Work in progress WebGPU 1.0.
```
I expect clang-format will complain.

Kangz replied:
Done

dneto0 commented on test/val/val_decoration_test.cpp:
```diff
@@ -12,7 +12,7 @@
 // limitations under the License.
 
-// Common validation fixtures for unit tests
+// Validation tests for decorations
```
Yay. Thanks.

Kangz replied:
Ack

dneto0 commented on test/val/val_fixtures.h:
```diff
@@ -43,8 +45,7 @@ class ValidateBase : public ::testing::Test,
 
-  // Performs validation on the SPIR-V code and compares the result of the
-  // spvValidate function
+  // Performs validation on the SPIR-V code.
```
Thank you for fixing this.

Kangz replied:
Ack

dneto0 commented on test/val/val_webgpu_test.cpp:
```diff
@@ -0,0 +1,43 @@
+  EXPECT_EQ(SPV_SUCCESS, ValidateInstructions(SPV_ENV_UNIVERSAL_1_3));
+
+  // Control case: OpUndef is disallowed in the WebGPU env
+  EXPECT_EQ(SPV_ERROR_INVALID_BINARY, ValidateInstructions(SPV_ENV_WEBGPU_0));
```
You should add a check for the result of the diagnostic string.
E.g. see uses of HasSubstr in val_capability_test.cpp.  For example the checks I just added for SubgroupShuffleINTEL

Kangz replied:
Done

Kangz commented:
PTAL again, addressed comments, some more additional commenting of existing code snuck in, and an additional commit to reduce the visibility of symbols in validate_instruction.cpp

Kangz commented:
Thanks for the review, the PR passed CI.

alan-baker commented on test/val/val_version_test.cpp:
```diff
@@ -112,6 +114,7 @@ INSTANTIATE_TEST_CASE_P(Universal, ValidateVersion,
     make_tuple(SPV_ENV_UNIVERSAL_1_1, SPV_ENV_OPENGL_4_2,    vulkan_spirv, false),
     make_tuple(SPV_ENV_UNIVERSAL_1_1, SPV_ENV_OPENGL_4_3,    vulkan_spirv, false),
     make_tuple(SPV_ENV_UNIVERSAL_1_1, SPV_ENV_OPENGL_4_5,    vulkan_spirv, false),
+    make_tuple(SPV_ENV_UNIVERSAL_1_0, SPV_ENV_WEBGPU_0,      vulkan_spirv, true),
```
This case is duplicated 4 times. I think you meant to update the universal environment comparison.

Kangz replied:
Done

alan-baker commented:
This failure seems unrelated. Filed #1632 for it.

dneto0 commented:
Rebased, squashed, and pushed into master as  ba602c90
